### PR TITLE
Add new physics object

### PR DESCRIPTION
**Pull request by andrewpap22**
### Greetings, 

Hello there, so, based on our discussion, i tried to add the new physics object, `PlanarCaloCell` and although it is somewhat successful there are some little problems that need to be fixed in order to be able for the merge to be done! 

 1. As you can see on the screenshot below, the calorimeters show correctly the energy deposits but they are not placed correctly. What is needed to be done: Each time a new cell is being read from the file and added to the event-display it needs to transform itself automatically, from the original default position to each position given by the plane from the .json file. Even though I tried to do that, I was not able to make it work and I would kindly request for your help on that. 
 2. And the second not so critical problem is that, I'm not quite sure how to load the new .json file automatically (by default) on the event display, so that I don't have to import it by hand every time through the GUI. As I saw in the code, the LHCbLoader is being called and not the PhoenixLoader on the following file, `packages/phoenix-ng/projects/phoenix-app/src/app/sections/lhcb/lhcb.component.ts` and thus because the new file is following the Phoenix format is not being read properly. 
 
pic-selected-210510-1731-49

One little note here, as you will see in the PhoenixLoader, where I add the PlanarCaloCell object, I'm not using the `this.addObjectType` because this method only accepts one parameter for the physics object, and in our case because we didn't want to make the PlanarCaloCell LHCb specific but globally experiment specific, there is a need for a second parameter in the `getPlanarCaloCell` method and that is the `plane` parameter that will get `plane` property from our .json file so we can get the corresponding plane position and then make the transform for each calorimeter based on the position of the plane.

I'm also tagging at this point Dr. @sponce in case I missinformed you with anything I wrote above. 

Thank you for your valuable help and time!

### CODE REVIEW

**Comment by sponce** [comment]
the problem is to transform the local (x,y) coordinates in the plane defined by (x,y,z,d) into global coordinates.
I believe you can achieve it by first creating the box in the z direction, then moving it by d along z and finally apply a rotation that transforms vector (0,0,1), i.e. z axis, intot (x,y,z) . This transformation can be defined via a quaternion via the setFromUnitVectors method (see https://discourse.threejs.org/t/get-euler-between-two-vector3s/5016)

**Comment by EdwardMoyse** [comment]
For your second question, I think it should be enough to change this line:
https://github.com/HSF/phoenix/blob/0783e224789b6dd8d07463a40835bcae99ea065e/packages/phoenix-ng/projects/phoenix-app/src/app/sections/lhcb/lhcb.component.ts#L58

To this:
```
this.eventDisplay.buildEventDataFromJSON(eventData);
```

The point is that `loadEventData` converts from LHCb format (using that loader) to Phoenix format. Since you're already in Phoenix format, you don't need this step.

File: packages/phoenix-ng/projects/phoenix-app/src/app/sections/lhcb/lhcb.component.ts [hunk 1]
```diff
@@ -52,7 +52,7 @@ export class LHCbComponent implements OnInit {
       'LHCb detector'
     );
 
-    fetch('assets/files/lhcb/00191749_0005296728.json')
+    fetch('assets/files/lhcb/LHCbEventDataV2.json')
       .then((res) => res.json())
       .then((eventData) => {
         this.loadEventData(eventData);
```
[the remark this answers is not in the document]
Comment (9inpachi) [reply packages/phoenix-ng/projects/phoenix-app/src/app/sections/lhcb/lhcb.component.ts]: Oh yes, indeed the method is in the `EventDisplay` class. :D

**Comment by andrewpap22** [comment]
### Hello there, 

So, as you can see in the picture below, we're successful now into displaying properly Planar Calorimeters! ( it should work for a general use case of planar calorimeters as well and not only for LHCb! )

pic-selected-210511-1857-56

One more question I have here is, based now on how I wrote the code on the `packages/phoenix-event-display/src/loaders/phoenix-loader.ts` for the `PlanarCaloCells` I'm not quite sure how I might add a new Cut. For example it would be nice to add a new Cut for the `energy` deposits. So I would like to ask again for your valuable help on this! Thank you in advance! 

Best regards, 

Andreas.

**Comment by andrewpap22** [comment]
P.S. You might notice that I made an `outterBox` on the `packages/phoenix-event-display/src/loaders/objects/phoenix-objects.ts` inside the `getPlanarCaloCell` method. I've added some comments reasoning why, but basically the better explanation is give on the following stackOverflow post: 

https://stackoverflow.com/questions/7541412/rotate-then-translate-in-three-js

**Comment by 9inpachi** [comment]
> One more question I have here is, based now on how I wrote the code on the packages/phoenix-event-display/src/loaders/phoenix-loader.ts for the PlanarCaloCells I'm not quite sure how I might add a new Cut. For example it would be nice to add a new Cut for the energy deposits. So I would like to ask again for your valuable help on this! Thank you in advance!

The `Cut` strategy is simple. The loader looks for the specified parameter in objects directly children of the collection and hides them based on the value range. For example consider the following cuts.

```ts
const cuts = [
  new Cut('energy', 0, 10000),
]
```

When you assign the physics object parameters to a 3D object like [here](https://github.com/HSF/phoenix/blob/f41920b7cb3b88113b02309f2c72463f25d0978f/packages/phoenix-event-display/src/loaders/objects/phoenix-objects.ts#L424) in your case, this means that Phoenix will look for the `energy` property in the `userData` of the Physics object and if the `energy` in that `userData` aka the value from the `caloCells` variable (`caloCells.energy`) is not with in the range which is `0` to `1000` in this example, it will hide it.

Since you are not using the `PhoenixLoader.addObjectType` or passing the cut argument to `this.ui.addCollection`. You are not leveraging this feature. ;)

In your current implementation, you could just uncomment the cuts and pass them to `this.ui.addCollection` and ideally it should work.

Apart from this, you could make use of the `PhoenixLoader.addObjectType` by refactoring your code like below:

```ts
if (eventData.PlanarCaloCells) {
  // (Optional) Cuts can be added to any physics object.
  const cuts = [
    new Cut('energy', 0, 10000)
  ];

  const addPlanarCaloCellsOptions = (
    typeFolder: GUI,
    typeFolderPM: PhoenixMenuNode
  ) => {
    const scalePlanarCaloCells = (value: number) => {
      this.graphicsLibrary
        .getSceneManager()
        .scaleChildObjects('PlanarCaloCells', value / 100, 'z');
    };

    if (typeFolder) {
      const sizeMenu = typeFolder
        .add({ PlanarCaloCellsScale: 100 }, 'PlanarCaloCellsScale', 1, 400)
        .name('PlanarCaloCells Size (%)');
      sizeMenu.onChange(scalePlanarCaloCells);
    }

    if (typeFolderPM) {
      typeFolderPM.addConfig('slider', {
        label: 'PlanarCaloCells Size (%)',
        value: 100,
        min: 1,
        max: 400,
        allowCustomValue: true,
        onChange: scalePlanarCaloCells,
      });
    }
  };

  this.getObjectTypeCollections(eventData.PlanarCaloCells).forEach(
    (collectionName) => {
      const plane = eventData.PlanarCaloCells[collectionName]['plane'];
      const unitVector = new Vector3(...plane.slice(0, 3)).normalize();
      eventData.PlanarCaloCells[collectionName]['cells'].forEach(
        (cell) => (cell['plane'] = [...unitVector.toArray(), plane[3]])
      );
    }
  );

  this.addObjectType(
    eventData.PlanarCaloCells,
    PhoenixObjects.getPlanarCaloCell,
    'PlanarCaloCells',
    false,
    cuts,
    addPlanarCaloCellsOptions
  );
}
```

**Comment by andrewpap22** [comment]
> > One more question I have here is, based now on how I wrote the code on the packages/phoenix-event-display/src/loaders/phoenix-loader.ts for the PlanarCaloCells I'm not quite sure how I might add a new Cut. For example it would be nice to add a new Cut for the energy deposits. So I would like to ask again for your valuable help on this! Thank you in advance!
> 
> The `Cut` strategy is simple. The loader looks for the specified parameter in objects directly children of the collection and hides them based on the value range. For example consider the following cuts.
> 
> ```ts
> const cuts = [
>   new Cut('energy', 0, 10000),
> ]
> ```
> 
> When you assign the physics object parameters to a 3D object like [here](https://github.com/HSF/phoenix/blob/f41920b7cb3b88113b02309f2c72463f25d0978f/packages/phoenix-event-display/src/loaders/objects/phoenix-objects.ts#L424) in your case, this means that Phoenix will look for the `energy` property in the `userData` of the Physics object and if the `energy` in that `userData` aka the value from the `caloCells` variable (`caloCells.energy`) is not with in the range which is `0` to `1000` in this example, it will hide it.
> 
> Since you are not using the `PhoenixLoader.addObjectType` or passing the cut argument to `this.ui.addCollection`. You are not leveraging this feature. ;)
> 
> In your current implementation, you could just uncomment the cuts and pass them to `this.ui.addCollection` and ideally it should work.
> 
> Apart from this, you could make use of the `PhoenixLoader.addObjectType` by refactoring your code like below:
> (If you want to try it out yourself, don't look at the code below ;))
> 
> ```ts
> if (eventData.PlanarCaloCells) {
>   // (Optional) Cuts can be added to any physics object.
>   const cuts = [
>     new Cut('energy', 0, 10000)
>   ];
> 
>   const addPlanarCaloCellsOptions = (
>     typeFolder: GUI,
>     typeFolderPM: PhoenixMenuNode
>   ) => {
>     const scalePlanarCaloCells = (value: number) => {
>       this.graphicsLibrary
>         .getSceneManager()
>         .scaleChildObjects('PlanarCaloCells', value / 100, 'z');
>     };
> 
>     if (typeFolder) {
>       const sizeMenu = typeFolder
>         .add({ PlanarCaloCellsScale: 100 }, 'PlanarCaloCellsScale', 1, 400)
>         .name('PlanarCaloCells Size (%)');
>       sizeMenu.onChange(scalePlanarCaloCells);
>     }
> 
>     if (typeFolderPM) {
>       typeFolderPM.addConfig('slider', {
>         label: 'PlanarCaloCells Size (%)',
>         value: 100,
>         min: 1,
>         max: 400,
>         allowCustomValue: true,
>         onChange: scalePlanarCaloCells,
>       });
>     }
>   };
> 
>   this.getObjectTypeCollections(eventData.PlanarCaloCells).forEach(
>     (collectionName) => {
>       const plane = eventData.PlanarCaloCells[collectionName]['plane'];
>       const unitVector = new Vector3(...plane.slice(0, 3)).normalize();
>       eventData.PlanarCaloCells[collectionName]['cells'].forEach(
>         (cell) => (cell['plane'] = [...unitVector.toArray(), plane[3]])
>       );
>     }
>   );
> 
>   this.addObjectType(
>     eventData.PlanarCaloCells,
>     PhoenixObjects.getPlanarCaloCell,
>     'PlanarCaloCells',
>     false,
>     cuts,
>     addPlanarCaloCellsOptions
>   );
> }
> ```

### Hello dear Fawad, 

Thank you very much for your suggestions and your knowledge! 

I committed the changes based on your suggestions and everything is working properly! I think now it's ready for the merge! 😃 

As for the refactor on the `phoenix-loader` that you mentioned, to make use of the `PhoenixLoader.addObjectType` I've tried this before as well myself and I was getting the following error: ( and still getting the same one after your suggestion ) 

pic-selected-210512-1256-20

### All the best, 

Andreas

**Comment by 9inpachi** [comment]
@andrewpap22 I think I get why. Because in the event data there is `plane` at the same level as a collection. So it reads `plane` as a collection tries to look for objects in it which is wrong.

May be doing this before passing data to `addObjectType` would fix it:

```ts
delete eventData.PlanarCaloCells[collectionName]['plane'];
```

Again, not sure if this is the problem. I can check after work today.

**Comment by EdwardMoyse** [comment]
Let's refactor in a new PR

**Pull request merged**